Add in the ability to control via parameters

clalancette opened the pull request:
This PR aims to allow the teleop_twist_joy node to control the buttons being used for various things via parameters.  The commits roughly do the following:

1.  Rewrite the `TeleopTwistJoy` class to inherit from Node, which makes further changes easier.
1.  Synchronizes the ros2 branch a bit with upstream.
1.  Sets the parameters on the node.
1.  Implements param_change_callback to make sure parameter changes get synchronized from the parameters into the `Impl` class.

That last point deserves a bit more explanation.  The design choice I had was to either keep the `Impl` class as-is and synchronize changes from the parameters to it, or to make the Parameters the "backing store" for the axes and scales.  I ended up keeping things closer to how they are today, mostly to minimize change from upstream.

Finally, here is a diff between this PR and https://github.com/ros-teleop/teleop_twist_joy/commit/10d79bebbf112f95f32713895f7cd0668faa83fc: https://gist.github.com/clalancette/334c53df8d80c2cc820c75595243cfda

clalancette committed:
Rewrite TeleopTwistJoy to inherit from Node.

This is the recommended way to do it, and will make adding
parameter events more straightforward.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Make teleop_twist_joy closer to upstream.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Rewrite parameter stuff to get the parameters using ROS 2 APIs.

This allows the node to be configurable like the ROS 1 version
was.  This also incorporates changes from
https://github.com/ros-teleop/teleop_twist_joy/pull/30, which
ensures that the node does not crash when invalid joystick
messages are published.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette commented:
CI:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_turtlebot-demo_linux&build=139)](http://ci.ros2.org/job/ci_turtlebot-demo_linux/139/)

mikaelarguedas commented on src/teleop_twist_joy.cpp:
```diff
@@ -68,63 +68,77 @@ struct TeleopTwistJoy::Impl
-  pimpl_->node = node;
-
-  pimpl_->cmd_vel_pub = node->create_publisher<geometry_msgs::msg::Twist>("cmd_vel",
+  pimpl_->cmd_vel_pub = this->create_publisher<geometry_msgs::msg::Twist>("cmd_vel",
```
Bringing that up again as I think that in general publishers (especially for command topics) should offer reliable reliability and only use best effort when subscriber request it

wjwwood replied:
+1

clalancette replied:
Makes sense to me, I'll change it.

mikaelarguedas commented on src/teleop_twist_joy.cpp:
```diff
@@ -68,63 +68,77 @@ struct TeleopTwistJoy::Impl
+    {
+      rcl_interfaces::msg::SetParametersResult result;
+      result.successful = true;
+      result.reason = "Success";
```
Nit: this could stay an empty string as it doesn't bring additional information while adding network overhead

clalancette replied:
I guess since this is parameters, it won't happen very often, so I'm not that concerned about it.  But I don't feel strongly about it, so I'll just make it empty.

mikaelarguedas commented on src/teleop_twist_joy.cpp:
```diff
@@ -68,63 +68,77 @@ struct TeleopTwistJoy::Impl
+      for (const auto& param : parameters) {
+        try {
+          std::string name = param.get_name();
+          auto int_it = std::find_if(int_parameters.begin(), int_parameters.end(), [&name](const JoyParameter<int>& j) -> bool { return name == j.name; });
```
Nit: this looks very long, same below

clalancette replied:
Yeah, good point, I've formatted it a bit nicer now.

mikaelarguedas commented on src/teleop_twist_joy.cpp:
```diff
@@ -68,63 +68,77 @@ struct TeleopTwistJoy::Impl
+                throw std::runtime_error("Cannot delete required parameter");
+              }
+            }
+            double_it->variable = param.as_double();
```
This means that any non double or known int parameter will be rejected. Is it intended?
e.g. this means that setting use_sim_time will now be impossible

clalancette replied:
No, that was just a bug, should be fixed now.

mikaelarguedas commented on src/teleop_twist_joy.cpp:
```diff
@@ -150,18 +164,16 @@ TeleopTwistJoy::TeleopTwistJoy(rclcpp::Node::SharedPtr & node)
+
 void TeleopTwistJoy::Impl::joyCallback(const sensor_msgs::msg::Joy::SharedPtr joy_msg)
 {
+  // Initializes with zeros by default.
```
Nit: As this is the same behavior as in ROS 1 I don't think it deserves a comment

clalancette replied:
That's only there to reduce the differences to upstream.

mikaelarguedas replied:
oh I missed it was present upstream. All good then :+1:

mikaelarguedas reviewed (commented):
Could you clarify the current use case?
Is there a requirement to enforce parameter existence at the moment ?

I'm asking because I see that as better fitted in the work started in rclcpp to declare and enforce parameter existence rather than starting implementing custom storage and logic in each leaf package using parameters.

Also the current state prevents setting other parameters that I think is not desired. A few more comments below

clalancette commented:
> Could you clarify the current use case?
> Is there a requirement to enforce parameter existence at the moment ?

The current requirement is to be able to change the axes used for control at both startup time and runtime.  Enforcing the parameter existence was something I tacked on since it seemed to make sense; deleting the basic parameters here would just cause weirdness down the line.

> 
> I'm asking because I see that as better fitted in the work started in rclcpp to declare and enforce parameter existence rather than starting implementing custom storage and logic in each leaf package using parameters.

Right, understood.  If this is on the roadmap, I can remove the piece where we enforce no deletion from this PR for now.

> 
> Also the current state prevents setting other parameters that I think is not desired. A few more comments below

Oops, that was a mistake on my part when I refactored, thanks for catching it.  I've fixed it now.

mikaelarguedas commented on src/teleop_twist_joy.cpp:
```diff
@@ -106,28 +106,24 @@ TeleopTwistJoy::TeleopTwistJoy() : Node("teleop_twist_joy_node")
       rcl_interfaces::msg::SetParametersResult result;
       result.successful = true;
-      result.reason = "Success";
+      result.reason = "";
```
This line can be removed

mikaelarguedas commented on src/teleop_twist_joy.cpp:
```diff
@@ -68,63 +68,73 @@ struct TeleopTwistJoy::Impl
+    {
+      rcl_interfaces::msg::SetParametersResult result;
+      result.successful = true;
+      result.reason = "";
```
oops commented on the wrong commit: redo of "this line can be removed"

mikaelarguedas commented on src/teleop_twist_joy.cpp:
```diff
@@ -68,63 +68,73 @@ struct TeleopTwistJoy::Impl
+      rcl_interfaces::msg::SetParametersResult result;
+      result.successful = true;
+      result.reason = "";
+      for (const auto& param : parameters) {
```
missing space before `&`, same multiple times below

clalancette replied:
Fixed now.

mikaelarguedas commented on src/teleop_twist_joy.cpp:
```diff
@@ -68,63 +68,73 @@ struct TeleopTwistJoy::Impl
+      for (const auto& param : parameters) {
+        try {
+          std::string name = param.get_name();
+          auto int_it = std::find_if(int_parameters.begin(), int_parameters.end(),
```
use 2 space indent, same below

clalancette replied:
Fixed now.

mikaelarguedas commented on src/teleop_twist_joy.cpp:
```diff
@@ -68,63 +68,73 @@ struct TeleopTwistJoy::Impl
-    node->get_parameter("scale_linear", pimpl_->scale_linear_map["x"]);
-    pimpl_->scale_linear_turbo_map["x"] = 1.0;
-    node->get_parameter("scale_linear_turbo", pimpl_->scale_linear_turbo_map["x"]);
+  // TODO(clalancette): node->get_parameter(s) doesn't support getting a map of
```
Can you provide the link of the ticket requesting the feature in the comment here?

clalancette replied:
Good idea, I just opened an issue and linked to it.

wjwwood commented on src/teleop_twist_joy.cpp:
```diff
@@ -68,63 +68,73 @@ struct TeleopTwistJoy::Impl
+
+  for(const auto& p : int_parameters) {
+    this->set_parameter_if_not_set(p.name, p.default_value);
+    this->get_parameter(p.name, p.variable);
```
I'm not convinced this is easier to read...

You could convert the initialization code into one liners with `node->get_parameter_or(...)` (http://docs.ros2.org/bouncy/api/rclcpp/classrclcpp_1_1_node.html#ae5d51004e6e4fc32194251587eb8711a), e.g.:

```c++
this->get_parameter_or("axis_linear", pimpl_->axis_linear_map["x"], 5);
this->get_parameter_or("scale_linear", pimpl_->scale_linear_map["x"], 0.5);
this->get_parameter_or("scale_linear_turbo", pimpl_->scale_linear_turbo_map["x"], 1.0);
// ...
```

clalancette replied:
> You could convert the initialization code into one liners with node->get_parameter_or(...)

I'm not sure that is true.  I really want the parameters to be set on the node; the documentation for `get_parameter_or` specifically says:

```
In all cases, the parameter remains not set after this function is called.
```

That's why I have the two-step `set_parameter_if_not_set` followed by `get_parameter`.  I might argue that this is a common use-case and we should have a helper that does this, but I didn't see one in the current API.

wjwwood replied:
Ah, you're right, it doesn't set the parameter to the default. But I guess I was assuming that you were getting the parameter to use locally, I didn't know you wanted to set it also. 👍

I still think I'd prefer to see it without a for loop, just expanded into lots of pairs of function calls, though that's more subjective.

mikaelarguedas commented on src/teleop_twist_joy.cpp:
```diff
@@ -68,63 +68,73 @@ struct TeleopTwistJoy::Impl
+                                       return name == j.name;
+                                     });
+          if (int_it != int_parameters.end()) {
+            int_it->variable = param.as_int();
```
Should this print the new value like it's done when the node starts? (same below)

clalancette replied:
That's a good idea, I'll move the printing to a method on the `Impl` class.

mikaelarguedas commented on src/teleop_twist_joy.cpp:
```diff
@@ -68,63 +68,73 @@ struct TeleopTwistJoy::Impl
+
+  for(const auto& p : int_parameters) {
+    this->set_parameter_if_not_set(p.name, p.default_value);
+    this->get_parameter(p.name, p.variable);
```
Could this use `get_parameter_or` ?

clalancette replied:
See https://github.com/ros2/teleop_twist_joy/pull/8#discussion_r207889563; unless there is something I'm missing, I don't think this will accomplish the same thing.

mikaelarguedas reviewed (commented):
> Enforcing the parameter existence was something I tacked on since it seemed to make sense; deleting the basic parameters here would just cause weirdness down the line. 
> Right, understood. If this is on the roadmap, I can remove the piece where we enforce no deletion from this PR for now.

We can go either way. As the ROS 1 code doesn't have such protection and has been under heavy use for years I don't feel like that's something that needs to be tackled here (as this is not part of the requirement).
On the other hand that's a small change that's useful so we can definitely just keep it until the feature is available at a lower level :+1:. So let's keep it here for now.

clalancette commented:
Thanks for the reviews!  I believe I've addressed all of the comments, so this is ready for another pass.

dirk-thomas commented on src/teleop_twist_joy.cpp:
```diff
@@ -64,104 +64,121 @@ struct TeleopTwistJoy::Impl
-    pimpl_->scale_angular_turbo_map["yaw"] = pimpl_->scale_angular_map["yaw"];
-    node->get_parameter("scale_angular_turbo", pimpl_->scale_angular_turbo_map["yaw"]);
+  // TODO(clalancette): node->get_parameter(s) doesn't support getting a map of
+  // values yet.  https://github.com/ros2/rclcpp/issues/529
```
I don't think this should call `get_parameter` for each parameter. Either call the existing function `get_parameters` to retrieve all parameters at once or implement a variation of the function returning a map if that is more convenient (as described in the referenced ticket).

clalancette replied:
The outcome of this is probably going to depend on what we decide for setting the default parameters or not.  I will say that I probably will not have time in the near future to implement the map variant on the node API; I'd rather see us get this in sooner with the existing TODO, and then fix it up once somebody has time to make that API.

dirk-thomas replied:
> I will say that I probably will not have time in the near future to implement the map variant on the node API

Implementing this API on top of the existing `get_parameters` method is literally taking 5 minutes. Please go ahead and do that as well use that function here.

dirk-thomas commented on src/teleop_twist_joy.cpp:
```diff
@@ -64,104 +64,121 @@ struct TeleopTwistJoy::Impl
+  double_parameters.push_back({"scale_angular_turbo_yaw", 0.5, pimpl_->scale_angular_turbo_map["yaw"]});
+
+  for(const auto & p : int_parameters) {
+    this->set_parameter_if_not_set(p.name, p.default_value);
```
Before the parameter was not being set to the default value. The default value was only used "in the code". Keeping that semantic seems like a good idea since otherwise you can't distinguish if the parameter was original set to a value or if it only got set from the default in code.

clalancette replied:
True, this is changing the semantic, but I don't think the existing semantic is a good one.  If you start up the node with the existing semantic and no parameters, then 0 parameters will be exposed (that is, `ros2 param list` will return an empty list for this node).  The only way to know that there are parameters that could be set is to read the code, and that just doesn't seem to make much sense.  I'd much rather have the parameters available at all times (and set to their default values at startup), at which point it is discoverable.

> Keeping that semantic seems like a good idea since otherwise you can't distinguish if the parameter was original set to a value or if it only got set from the default in code.

It seems to me that what mostly matters is what the current value for each parameter is (it is not like this is a complex node, after all).  Is there a particular reason that you see where distinguishing this matters?

dirk-thomas replied:
I am ok with consciously changing the semantic. Making the default value introspectable is valuable indeed. :+1:

wjwwood reviewed (dismissed):
I don't have any blocking comments any more, so lgtm! But I'd recommend letting Dirk and Mikael finish their reviews too.

mikaelarguedas commented:
Same here, but I'll leave it up to @clalancette and @dirk-thomas to address the last comments and give final approval

clalancette commented:
With ros2/rclcpp#575 being merged, we can now revisit this.  I've cleaned up this PR to collapse the commits together, use the new `get_parameters` rclcpp APIs, and incorporate changes that are pending in upstream ros-teleop/teleop_twist_joy#30.  With all of that, I think this is ready to be re-reviewed, so I'm going to dismiss @wjwwood's review (the code has changed significantly since then), and put this back in "in-review".

clalancette commented:
@ros2/team I'd like to get this into the [Crystal patch release 2](https://github.com/ros2/ros2/issues/647), so review is appreciated.

jacobperron commented on src/teleop_twist_joy.cpp:
```diff
@@ -46,188 +46,156 @@ namespace teleop_twist_joy
-    rmw_qos_profile_sensor_data);
-  pimpl_->joy_sub = node->create_subscription<sensor_msgs::msg::Joy>("joy",
+  pimpl_->cmd_vel_pub = this->create_publisher<geometry_msgs::msg::Twist>("cmd_vel",
+    rmw_qos_profile_default);
```
nit: Since this is the default value for the argument, we can omit the argument

clalancette replied:
While that is true, I think it is better to keep it for documentation purposes.  What I've noticed while working on some of my robot projects is that mis-matched publishers/subscribers is an issue that continually trips people up, so having it explicit here seems better to me.  I'm going to leave it for now as-is.

jacobperron commented on src/teleop_twist_joy.cpp:
```diff
@@ -46,188 +46,156 @@ namespace teleop_twist_joy
-  //   nh_param->getParam("scale_linear_turbo", pimpl_->scale_linear_turbo_map);
-  // }
-  // else
+    rmw_qos_profile_default);
```
nit: remove default argument

clalancette replied:
Ditto.

jacobperron commented on src/teleop_twist_joy.cpp:
```diff
@@ -46,188 +46,156 @@ namespace teleop_twist_joy
-  //   nh_param->getParam("scale_angular_turbo", pimpl_->scale_angular_turbo_map);
-  // }
-  // else
+  this->get_parameters("axis_angular", pimpl_->axis_angular_map);
```
Since we want default values, instead of using `get_parameters()` and checking if values are set, could we do something like this instead?

```c++
this->get_parameter_or_set("axis_angular.yaw", pimpl_->axis_angular_map["yaw"], 2);
```
This way we wouldn't need all of the `if` blocks.

clalancette replied:
The problem is the way in which these are accessed.  Just looking at `axis_angular`, for instance, the rules are generally:

1.  If `roll` or `pitch` are specified in the parameters when the node is launched, set them in the `axis_angular_map`.  Otherwise, they don't exist.
2.  If `yaw` is specified in the parameters when the node is launched, set them in `axis_angular_map`.  If unspecified, set it anyway to the default value of 5.
3.  When a joystick message arrives, lookup into `axis_angular_map` for each of `roll`, `pitch`, `yaw` to get their axis.  If it doesn't exist in the map, then just fill in 0.0 in the corresponding field in the outgoing `cmd_vel` message.

Thus, to implement what you are proposing, we'd have to change the logic for axes lookup to look for a default value (say, -1), instead of just whether it exists in the map.  While I'm not fundamentally opposed to doing that, it is a departure from upstream https://github.com/ros-teleop/teleop_twist_joy/blob/indigo-devel/src/teleop_twist_joy.cpp#L74 , so I think I'd prefer to leave this as-is for now.  We can always revisit later (both here and upstream), since it isn't an API change, just an implementation change.

jacobperron replied:
I see. But, I think the following produces the same outcome:

```diff
this->get_parameters("axis_angular", pimpl_->axis_angular_map);
+this->get_parameter_or_set("axis_angular.yaw", pimpl_->axis_angular_map["yaw"], 2);
-if (pimpl_->axis_angular_map.count("yaw") == 0)
-{
-  pimpl_->axis_angular_map["yaw"] = 2;
-  this->set_parameter_if_not_set("axis_angular.yaw", pimpl_->axis_angular_map["yaw"]);
-}
```

You can leave it as-is though :)

jacobperron commented on src/teleop_twist_joy.cpp:
```diff
@@ -46,188 +46,156 @@ namespace teleop_twist_joy
+  if (pimpl_->scale_linear_map["normal"].count("x") == 0)
+  {
+    pimpl_->scale_linear_map["normal"]["x"] = 0.5;
+    this->set_parameter_if_not_set("scale_linear.x", pimpl_->scale_linear_map["normal"]["x"]);
```
Similarly, this block could be replaced by:

```c++
this->get_parameter_or_set("scale_linear.x", pimpl_->scale_linear_map["normal"]["x"], 0.5);
```

jacobperron commented on src/teleop_twist_joy.cpp:
```diff
@@ -46,188 +46,156 @@ namespace teleop_twist_joy
-  pimpl_ = new Impl();
-
-  pimpl_->node = node;
+  pimpl_ = new Impl;
```
Consider using smart pointer instead.

clalancette replied:
Yeah, I went back and forth on that.  I ended up leaving it as a raw pointer to be more compatible with upstream https://github.com/ros-teleop/teleop_twist_joy, which has to support all the way back to Indigo.  So I think I'll leave this one as-is as well.

jacobperron reviewed (approved):
LGTM

Couple nits and a suggestion, but if you prefer to leave it as-is, that's cool too.

clalancette commented:
@jacobperron Thanks for the review!  I decided not to take any of your comments for now, so I'm going to merge this after a quick run through CI.  If you still want to discuss one or more of them, let me know, I'm happy to do so.

clalancette commented:
CI:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6112)](http://ci.ros2.org/job/ci_linux/6112/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2587)](http://ci.ros2.org/job/ci_linux-aarch64/2587/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5041)](http://ci.ros2.org/job/ci_osx/5041/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=5982)](http://ci.ros2.org/job/ci_windows/5982/)

clalancette committed:
Switch to int64_t for integers.

This is because parameters in ROS 2 only come in int64_t
flavors, and MSVC complains about a loss of precision.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Make parameter unambiguous.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Change to ament_cmake to make Windows happy.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>
